### PR TITLE
Feature: Form Field types (PDF 1.7 Section 12.7.4)

### PR DESCRIPTION
**Pull request by tingerrr**
This PR adds the writers and types for writing various types form fields.

- [x] Text Fields
- [x] Choice Fields
- [x] Button Fields
- [ ] ~~Signature Fields~~ (Left out to reduce review surface to in turn ensure correct implementation)

### CODE REVIEW

**Comment by tingerrr** [comment]
I once again went with a prefix based approach to make the specific functions for `/V` and `/DV` exist in tandem. I don't like this solution all that much, we can probably clean this up in the future by providing specialized writers for cases where dictionaries have specific fields for a specific subtype.

**Comment by tingerrr** [comment]
While this is marked as ready for review, there is still some work to do, like exporting the types and such.

File: src/forms.rs [hunk 1]
```diff
@@ -76,8 +77,211 @@ impl<'a> Field<'a> {
...
+        self.dict.pair(Name(b"V"), value);
+        self
+    }
+
+    /// Start writing the `/DV` attribute to set the default value of this text
+    /// field. Only permissible on text fields.
+    pub fn text_default_value(&mut self, value: TextStr) -> &mut Self {
+        self.dict.pair(Name(b"DV"), value);
+        self
+    }
```
Comment (reknih) [inline src/forms.rs:104]: Where is `DV` defined for text fields?

File: src/forms.rs [hunk 2]
```diff
@@ -76,8 +77,211 @@ impl<'a> Field<'a> {
...
+/// Only permissible on fields containing variable text.
+impl<'a> Field<'a> {
+    /// Write the `/DA` attribute containing a sequence of valid page-content
+    /// graphics or text state operators that define such properties as the
+    /// field's text size and colour. Only permissible on fields containing
+    /// variable text.
+    pub fn vartext_default_appearance(&mut self, appearance: Str) -> &mut Self {
+        self.dict.pair(Name(b"DA"), appearance);
+        self
+    }
```
Comment (reknih) [inline src/forms.rs:116]: Would it be more convenient if this returned a writer for a content stream?

File: src/forms.rs [hunk 3]
```diff
@@ -76,8 +77,211 @@ impl<'a> Field<'a> {
...
+
+    /// Write the `/Q` attribute to set the quadding (justification) that shall
+    /// be used in dispalying the text. Only permissible on fields containing
+    /// variable text.
+    pub fn vartext_quadding(&mut self, quadding: Quadding) -> &mut Self {
+        self.dict.pair(Name(b"Q"), quadding as u32 as i32);
```
Comment (reknih) [inline src/forms.rs:122]: `as u32 as i32` ooff.
Can we drop one of these casts?

File: src/forms.rs [hunk 4]
```diff
@@ -76,8 +77,211 @@ impl<'a> Field<'a> {
...
+        self
+    }
```
[the remark this answers is not in the document]
Comment (reknih) [reply src/forms.rs:124]: I'll check with @laurmaedje

File: src/forms.rs [hunk 5]
```diff
@@ -119,5 +357,106 @@ bitflags::bitflags! {
...
+        /// value.
+        const PUSHBUTTON = 1 << 17;
+        /// A group of radio buttons within a radio button field that use the
+        /// same value for the on state will turn on and off in unison; that
+        /// is if one is checked, they are all checked. If unset, the buttons
+        /// are mutually exclusive (the same behavior as HTML radio buttons).
+        /// PDF 1.5+.
+        const RADIOS_IN_UNISON = 1 << 26;
+
+        // text field specific flags
```
Comment (reknih) [inline src/forms.rs:381]:
```suggestion
        // Text field specific flags
```

File: src/forms.rs [hunk 6]
```diff
@@ -119,5 +357,106 @@ bitflags::bitflags! {
         /// The field shall not be exported by a
         /// [submit-form](crate::types::ActionType::SubmitForm)[`Action`].
         const NO_EXPORT = 1 << 3;
+
+        // button specific flags
```
Comment (reknih) [inline src/forms.rs:361]:
```suggestion
        // Button specific flags
```

File: src/forms.rs [hunk 7]
```diff
@@ -119,5 +357,106 @@ bitflags::bitflags! {
...
+        /// and the text is layed out into these combs. May only be set if
+        /// the [`Field::max_len`] property is set and if the [`MULTILINE`],
+        /// [`PASSWORD`] and [`FILE_SELECT`] flags are clear. PDF 1.5+.
+        const COMB = 1 << 25;
+        /// The value of this field shall be a rich text string. If the field
+        /// has a value, the [`TextField::rich_text_value`] shall specify the
+        /// rich text string. PDF 1.5+.
+        const RICH_TEXT = 1 << 26;
+
+        // choice field specific flags
```
Comment (reknih) [inline src/forms.rs:413]:
```suggestion
        // Choice field specific flags
```

File: src/forms.rs [hunk 8]
```diff
@@ -119,5 +357,106 @@ bitflags::bitflags! {
...
+        const MULTILINE = 1 << 13;
+        /// The text contains a password and should not be echoed visibly to
+        /// the screen.
+        const PASSWORD = 1 << 14;
+        /// The entered text represents a path to a file who's contents shall be
+        /// submitted as the value of the field. PDF 1.4+.
+        const FILE_SELECT = 1 << 21;
+        /// The entered text shall not be spell-checked, can be used for text
+        /// and choice fields.
+        const DO_NOT_SPELL_CHECK = 1 << 23;
```
Comment (reknih) [inline src/forms.rs:394]: Maybe this should be in some shared section?

File: src/forms.rs [hunk 9]
```diff
@@ -119,5 +357,106 @@ bitflags::bitflags! {
...
+    }
+
+    /// Add an option with the given value and export value.
+    pub fn export(&mut self, value: TextStr, export_value: TextStr) -> &mut Self {
+        self.array.push().array().items([export_value, value]);
+        self
     }
 }
+
+deref!('a, ChoiceOptions<'a> => Array<'a>, array);
```
Comment (reknih) [inline src/forms.rs:462]: Should be moved directly below the Choice impl

File: src/forms.rs [hunk 10]
```diff
@@ -76,8 +77,211 @@ impl<'a> Field<'a> {
...
+    pub fn choice_default_values<'b>(
+        &mut self,
+        options: impl IntoIterator<Item = TextStr<'b>>,
+    ) -> &mut Self {
+        self.dict.insert(Name(b"DV")).array().items(options);
+        self
+    }
+}
+
+/// Only permissible on button fields.
```
Comment (reknih) [inline src/forms.rs:219]: Why are buttons below the other stuff when they come first in the spec?

**Review by reknih — COMMENTED** [review]
It's a huge section of the spec so there are bound to be some comments. You are making good progress.

File: src/forms.rs (hunk 1 above)
Comment (tingerrr) [reply src/forms.rs:104]: 12.7.5.3 Reset-Form Action (linked to from Section 12.7.3.1 (Table 220, entry `/DV`)) it says that the value in `/DV` is the value `/V` reverts to when a `Reset-Form` action is done. Therefore, it should be available on all fields in the same format as `/V` is.

File: src/forms.rs (hunk 10 above)
Comment (tingerrr) [reply src/forms.rs:219]: I'll rearrange it to fit the spec and the rest of the code base better.

File: src/forms.rs (hunk 2 above)
Comment (tingerrr) [reply src/forms.rs:116]: I honestly don't know, the spec simply said it's a string, so I figured `Str` would be it, is a writer for this kind of text operators available already, or should I add one?

Comment (reknih) [reply src/forms.rs:230]: Okay nvm its probably too complicated

File: src/forms.rs [hunk 11]
```diff
@@ -76,8 +77,211 @@ impl<'a> Field<'a> {
...
+impl<'a> Field<'a> {
+    // TODO: the spec likely means the equivalent of unicode graphemes here
+    //       for characters
+
+    /// Write the `/MaxLen` attribute to set the maximum length of the fields
+    /// text in characters. Only permissible on text fields.
+    pub fn text_max_len(&mut self, len: i32) -> &mut Self {
+        self.dict.pair(Name(b"MaxLen"), len);
+        self
+    }
```
[the remark this answers is not in the document]
Comment (tingerrr) [reply src/forms.rs]: @reknih this is still relevant, unfortunately moving the code around has marked this as outdated.

File: src/forms.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tingerrr) [reply src/forms.rs]: Since it was merged without any request for change I assume this is resolved.

**Comment by PavanSamyak** [comment]
I'm new to typst can anyone guide me how to add interactive checkbox to pdf without postprossing

**Pull request merged**